### ISSUE TITLE
C4: Typed conformance and frontend enablement Implementation Plan

### ISSUE THREAD

**Issue by JueonPark** [JueonPark/sym#144]
**Goal:** Prove typed relocation semantics end to end and enable frontend import only for compiler/runtime combinations that pass the declared conformance contract.

**Architecture:** Build independent scalar references and a versioned conformance corpus, then run compiler→artifact→bind→execution against them. Extend the existing optional Torch recipe/import/artifact layer behind explicit typed capability checks. Keep unsupported captures and execution variants on their original path with reason-coded diagnostics.

**Tech Stack:** LLVM/MLIR 21.1.8; standalone C++ runtime and existing CPU/CUDA kernels; CPython 3.14.7 standard GIL, PyTorch 2.14.0 CPU/cu126, pybind11 3.0.4, NumPy >=2.3.3,<3 and pytest >=9,<10. Record resolved versions and actual CUDA device evidence.

**Spec:** [#132](https://github.com/JueonPark/sym/issues/132), [project plan §2/C4](https://github.com/JueonPark/sym/blob/main/docs/project-finalization-plan.md#2-extend-plans-with-cast-quantize-and-dequantize), [#131](https://github.com/JueonPark/sym/issues/131), and the C1–C3 contracts.

## Global Constraints

- "Unsupported cases execute through the original PyTorch path with a recorded reason."
- "Keep torch dependencies outside the core runtime."
- The requested dtype and numerical semantics determine legal implementations. Optional cost advice cannot introduce quantization, skip a stage, or substitute an unimplemented kernel.
- Use independent scalar/Torch/NumPy results, not another production kernel or compiler index implementation, as the oracle.
- Exact layout and integer outputs remain exact. Floating comparisons follow C1's explicit policy; any dequantization tolerance is justified and fixed per operation before testing.
- Typed capture, successful compilation, successful binding, and enabled execution are separate support states.
- Existing no-grad, no-mutation, alias/layout, blocking/lifetime, symbolic-guard, and original-callable contracts remain applicable. A D2H request uses its forward typed computation.
- No int4 support, calibration/training, broad Torch-version matrix, or performance threshold is added.

---

**Proposed child title:** `[Finalize][Compiler][C4] Qualify typed execution and enable frontend import`

**Dependencies:** [C1](https://github.com/JueonPark/sym/issues/141) semantics, [C2](https://github.com/JueonPark/sym/issues/142) folding, [C3](https://github.com/JueonPark/sym/issues/143) typed artifacts/binding, and R3 typed dispatch in [#133](https://github.com/JueonPark/sym/issues/133). Reuse the completed T1/T2 frontend from #134/#135 and #140. Reference cases can be prepared after [C1](https://github.com/JueonPark/sym/issues/141); execution acceptance needs real R3. Automatic transfer replacement additionally requires T3 [#136](https://github.com/JueonPark/sym/issues/136) and R2 storage/stream handling; standalone typed conformance does not wait for those, and must not claim their completion. Coordinate parameter/weight freshness cases with T4 [#137](https://github.com/JueonPark/sym/issues/137) and final handoff R4.

**Compiler subissues:** [C1](https://github.com/JueonPark/sym/issues/141), [C2](https://github.com/JueonPark/sym/issues/142), [C3](https://github.com/JueonPark/sym/issues/143), [C4](https://github.com/JueonPark/sym/issues/144).

## Starting evidence

The current `reloc_torch.Recipe`, emitter, and artifact validator describe layout-only f32/f16/i8 plans with unchanged dtype. Import candidates retain original callables and the compiler is a later proof gate. The strict v0/schema-1 loader must remain compatible while C3 adds a typed branch. Existing raw float→int casts or quantized Tensor storage must not be silently reinterpreted as the new signed-i8 quantization contract.

## File map

| Action | Path | Responsibility |
| --- | --- | --- |
| Create | `libreloc/python/tests/typed_reference.py` | Independent scalar cast/quantize/dequantize and recipe replay oracle |
| Create | `libreloc/python/tests/test_typed_conformance.py` | Torch-free public compiler/runtime, footprint, and portability conformance using NumPy/pyreloc |
| Create | `libreloc/test/corpus/typed/` | Compiler-generated typed fixtures, exact parameters, policies, and expected outputs |
| Create | `test/dialect/reloc/typed_conformance.mlir` | Compiler structure/order/serialization witnesses |
| Consume | `libreloc/python/reloc_torch/recipe.py`, `mlir_emit.py` | C3's frozen typed records and deterministic emitter; extend only when an audited capture needs an explicitly supported representation |
| Modify | `libreloc/python/reloc_torch/compat.py`, `fx_import.py`, `eligibility.py` | Audited Torch targets, parameter provenance, and typed admission gates |
| Modify | `libreloc/python/reloc_torch/artifact.py`, `compiler.py` | Consume C3's typed contract with full recipe/parameter association |
| Create | `libreloc/python/tests/torch_frontend/test_typed_import.py` | Real FX capture, typed compile, fallback, parameter, and capability tests |
| Modify | `libreloc/test/QuantTest.cpp`, `CudaKernelsTest.cpp`, relevant R3 execution tests | Qualify selected scalar/SIMD/CUDA implementations against the same semantics |
| Create | `libreloc/python/examples/torch_typed_relocation.py`, `docs/typed-relocation-support.md` | Runnable explicit usage, support matrix, and reproduction evidence |
| Modify | `docs/torch-support.md`, `.github/workflows/build.yml` | Link current enablement and require real compiler/runtime acceptance in CI |

## Task 1: Define independent numerical references and corpus

**Interfaces:** Produce reference functions `cast_f32_f16`, `cast_f16_f32`, `quantize_f32_s8`, `dequantize_s8_f32`, and ordered recipe replay in `typed_reference.py`, implementing C1's published policy independently of production executors. Corpus metadata records that policy and exact parameter bits.

- [ ] Implement the oracle's stage arithmetic exactly as C1 specifies, including intermediate precision and parameter interpretation. Do not use `quantizePackF32S8`, the compiler's inverse map, or a production executor to generate expected answers.
- [ ] Pin finite cast bits and integer outputs with independent expected vectors. For NaNs, signed zero, subnormals, infinities, and overflow, assert the exact C1 promise rather than a broad `allclose` that hides a policy mismatch.
- [ ] State absolute/relative tolerance (or exact comparison where required) for dequantization before running kernel comparisons. Explain each allowance from C1's arithmetic contract; do not widen it after observing a failure.
- [ ] Include rounding ties, int8 limits, nonzero zero points, per-tensor/per-channel parameters, channel-axis permutation/split, constant pad, padded channel axes, and a lossy cast/quantize roundtrip. Unsupported kernel cases must have explicit capability expectations.
- [ ] Add the pad-order witness: with scale 0.5 and zero point 0, f32 fill 1.0 before quantization becomes code 2, whereas s8 fill code 1 after quantization remains 1.
- [ ] Generate typed plan bytes through the public exporter and save metadata/reference outputs separately. Repeated generation must be deterministic. Keep all existing layout-only corpus bytes and exact tests intact.
- [ ] Run the reference/corpus tests and commit them before enabling typed frontend paths.

## Task 2: Qualify the real compiler-to-runtime path

**Interfaces:** Consume C3's public exporter/decoder/binder and R3's real capability/preflight/dispatch API. Produce execution evidence identifying the requested stages, actual selected path, logical output metadata, and source/wire/destination/parameter bytes.

- [ ] Add failing end-to-end tests before integrating R3. A missing exporter, typed runtime, or required CPU reference path fails the acceptance test; a mocked dispatcher cannot satisfy this task.
- [ ] Run at least one mixed layout+cast and one layout+quantize/dequantize recipe through public compilation, save/load, symbol/parameter binding, and actual runtime execution. Compare to independent ordered replay, including shape, strides, dtype, offset, and values.
- [ ] Reuse one symbolic artifact across several valid shapes and parameter bindings. Reject invalid extents, divisibility, repeated-symbol equality, channel lengths, scales, zero points, and byte overflow before launch; assert that execution counters remain unchanged on rejection.
- [ ] Test the declared f32→s8→f32 path with its distinct source/wire/destination footprints, including a padded destination. Verify actual transferred tensor bytes and any parameter transfers are accounted for separately.
- [ ] Force the original CPU-transform path through R3's public API. Compare every automatically selected alternative only where R3 reports the same supported semantics. Missing calibration must not disable a correct reference path or authorize a different precision.
- [ ] Run signed-i8 and half edge cases through each claimed scalar/SIMD/CUDA implementation. Qualify H2D/D2H rows only with actual transfers; specifically verify that D2H does not use inverse-layout scatter for a non-invertible typed program.
- [ ] Reload artifacts in a fresh process and rebind external parameters. Change parameter values between completed calls and verify fresh results or an explicit invalidation; invalid updated scales must fail preflight.
- [ ] Commit runtime conformance before reporting any typed combination as executable.

## Task 3: Import only proved Torch value transformations

**Interfaces:** Consume C3's frozen cast/quantize/dequantize recipe records, emitter, and parameter declarations, and produce audited FX mappings into them. Preserve canonical identity, parameter/symbol provenance, portable admission, and the original callable's complete argument order.

- [ ] Capture the actual raw FX and ATen forms under Torch 2.14.0 before adding rules. Keep internal access in `compat.py`; do not copy historical target names or assume every quantization API yields ordinary `torch.int8` storage.
- [ ] Import ordinary f32/f16 casts only when their semantics match C1. Import quantization/dequantization through audited explicit operations or complete, proved decompositions; a float→int `.to()` is not evidence of saturating scale/zero-point quantization.
- [ ] Preserve externally supplied scale/zero-point operands and their logical channel axes. Do not read tensor data, `.item()`, `int(SymInt)`, or capture hints to specialize a recipe accidentally.
- [ ] Verify captured recipes use C3's canonical keys and artifact association, including stage order, policies, dtypes, exact inline values, and parameter declaration identities. Runtime parameter values remain validated bindings, not an unsafe pointer-based cache key.
- [ ] Test actual captured mixed-layout cast and quantization recipes, compiler bails, unknown numerical policies, unsupported versions, missing capabilities, gradients, mutation, aliases, and escaping intermediates. Original graph/callable behavior must remain intact on every exclusion.
- [ ] Gate import/compile/bind/execute states separately. Keep typed graph replacement excluded until the real T3/R2/R3 path qualifies that row; merely accepting typed IR cannot turn on automatic execution.
- [ ] Test fake metadata against real output for any custom-op route that is enabled with T3. Registration/opcheck is additional evidence, not a substitute for numerical comparison.
- [ ] Run the new typed frontend tests plus existing T1/T2 regression tests and commit the capability-gated import.

## Task 4: Publish the support matrix and reproducible handoff

**Interfaces:** Produce `docs/typed-relocation-support.md` with operation, source/destination/wire dtype, numerical policy, parameter mode, layout, direction, artifact version, runtime path, support state, reason, and a linked test for each row.

- [ ] Publish supported and explicitly unsupported rows, including nonzero-zero-point and channel-layout combinations that existing fast kernels cannot implement. Distinguish compiler-only, CPU-reference, and actual CUDA evidence.
- [ ] Add a runnable explicit example using generated plans and supplied parameters. Include mixed layout+cast and layout+quantize/dequantize, symbolic rebinding, output validation, and recorded path/byte counters.
- [ ] Make the CPU CI leg require the real exporter, typed decoder/binder, and R3 CPU execution path. Keep standalone runtime tests Torch-free. Record actual CUDA versions/device/pass-skip counts separately; skips cannot certify a claimed GPU row.
- [ ] Document which T3/R2 capabilities remain necessary for automatic transfer replacement and which T4/R4 work remains for weights/lifetime handoff. Keep performance measurements descriptive, with no speedup gate.
- [ ] Run the full layout and typed regressions, publish the exact reproduction commands, and commit support documentation with its executable example.

## Acceptance and handoff

- [ ] The existing layout corpus stays exact and v0 artifacts remain usable.
- [ ] Cast/quantization match the declared reference; dequantization meets its predeclared comparison rule, including numerical edge cases.
- [ ] At least one mixed layout+cast and one mixed layout+quantize/dequantize recipe reaches the real runtime through public artifacts and matches independent replay.
- [ ] Symbolic shape and parameter changes are safe; failed guards never launch typed execution.
- [ ] Frontend support is enabled only for jointly verified compiler/runtime combinations; unsupported captures retain original behavior and a reason.
- [ ] Support matrix, generated fixtures, fresh-process reuse, runnable examples, and CPU/actual-CUDA evidence are linked for the #132/#133 handoff.

After configuring a matching qualification build:

```bash
export SYM_BUILD="$PWD/build/torch-cpu"
export PYTHONPATH="$SYM_BUILD/python"
export SYM_OPT="$SYM_BUILD/sym/tools/sym-opt"
export SYM_RELOC_EXPORT="$SYM_BUILD/sym/tools/sym-reloc-export"
cmake --build "$SYM_BUILD" -j2
cmake --build "$SYM_BUILD" --target check-sym
ctest --test-dir "$SYM_BUILD" --output-on-failure
/tmp/sym-torch-cpu/bin/python -m pytest libreloc/python/tests -m 'not gpu' -q
```

For claimed CUDA rows, switch to the qualified cu126 environment and an R2/R3-enabled CUDA build, reset `PYTHONPATH`/tool paths, and run the new `gpu`-marked conformance/frontend cases on actual hardware. CPU or fake execution is not GPU acceptance.
